presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -963,7 +963,9 @@ else if (criteria instanceof JoinOn) {
                 // we do it further down when after we determine which subexpressions apply to left vs right tuple)
                 ExpressionAnalyzer analyzer = ExpressionAnalyzer.create(analysis, session, metadata, sqlParser, accessControl);
 
-                Type clauseType = analyzer.analyze(expression, output);
+                // need to register coercions in case when join criteria requires coercion (e.g. join on char(1) = char(2))
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, output);
```

Comment:
Good catch, Karol!
I've fixed it in `StatementAnalyzer.visitJoin` (to register coercions similarly to what I've added before for `clauseType`); 435bd492b65161157364c10d376cc9e280195635.
I've added a test case for that here: 24bd232752b69881e78a0d194450f2829e60ffdf.

However, I believe the problem is broader. Because now `canonicalizeExpression` is relying on outdated AST (because of potential coercions added in previous `analyzeExpression`), so it's not taking into consideration them, therefore the output of `canonicalizeExpression` may be incorrect. Let me think what would be the case to make it fail. But anyway, even if such a case does not exist now, it may occur in future when `canonicalizeExpression` changes...

@martint what do you think about extracting `canonicalizeExpression` to `PlanOptimizer`, similarly as I've done with `optimizeExpression`?